File: LeetCode/Medium/0328-odd-even-linked-list/0328-odd-even-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode oddEvenList(ListNode head) {

        if(head == null || head.next == null) return head;
        
        ListNode tempEvenHead = head.next;
        ListNode tempEven = tempEvenHead;
        ListNode tempOdd = head;

        while(true) {

            // 2개씩 넘어감
            if(tempOdd != null && tempOdd.next != null && tempOdd.next.next != null) {
                tempOdd.next = tempOdd.next.next;
                tempOdd = tempOdd.next;
            }
            else break;

            if(tempEven != null && tempEven.next != null && tempEven.next.next != null) {
                tempEven.next = tempEven.next.next;
                tempEven = tempEven.next;
            }
            else break;

        }

        tempEven.next = null;
        tempOdd.next = tempEvenHead;

        // printNode(head);

        return head;
    }

    private void printNode(ListNode node) {

        ListNode temp = node;

        while(temp != null) {
            System.out.print(temp.val + " / ");
            temp = temp.next;
        }

    }
}

File: LeetCode/Medium/0328-odd-even-linked-list/0328-odd-even-linked-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun oddEvenList(head: ListNode?): ListNode? {

        if(head == null || head.next == null) return head

        val tempEvenHead = head.next
        var tempEven = tempEvenHead
        var tempOdd = head

        while(true) {
            // 2개씩 넘어감
            if(tempOdd != null && tempOdd.next != null && tempOdd.next.next != null) {
                tempOdd.next = tempOdd.next.next
                tempOdd = tempOdd.next
            }
            else break;

            
            if(tempEven != null && tempEven.next != null && tempEven.next.next != null) {
                tempEven.next = tempEven.next.next
                tempEven = tempEven.next
            }
            else break;
        }

        if(tempEven != null) tempEven.next = null
        if(tempOdd != null) tempOdd.next = tempEvenHead
        
        return head
    }
}